### ISSUE TITLE
PC version lagging for me - Android version is not

### ISSUE THREAD

**Issue by nk33** [moonlight-stream/moonlight-qt#578]
I am getting lag and slow connection speeds on my PC client all the time now.

this happens locally and remotely and no matter what bitrate i use - even really low like 2mb....

This is not happening on my android cleints - they are fine.

I am using H264 accross the board

I have tested it on my laptop and pc - pc is connected via lan and laptop wifi - both same issue

**Comment by cgutman** [comment]
Please post your Moonlight logs (as mentioned in the issue template).

**Comment by nk33** [comment]
sorry mate

please find atatched
[Moonlight-1618827849.log](https://github.com/moonlight-stream/moonlight-qt/files/6336964/Moonlight-1618827849.log)
[Moonlight-1618827791.log](https://github.com/moonlight-stream/moonlight-qt/files/6336965/Moonlight-1618827791.log)
[Moonlight-1618827773.log](https://github.com/moonlight-stream/moonlight-qt/files/6336966/Moonlight-1618827773.log)
[Moonlight-1618827728.log](https://github.com/moonlight-stream/moonlight-qt/files/6336967/Moonlight-1618827728.log)

**Comment by nk33** [comment]
Guys would you like any more info for this issue? it happens from any PC remotely - lag. Android device on same network, no lag

**Comment by cgutman** [comment]
The logs confirm what I had guessed. You're losing just slightly more packets than FEC can recover at the standard 20% FEC level. 

```
00:00:27 - SDL Info (0): Unrecoverable frame 670: 7+3=10 received < 11 needed
00:00:28 - SDL Info (0): Unrecoverable frame 687: 7+3=10 received < 11 needed
00:00:29 - SDL Info (0): Unrecoverable frame 780: 7+3=10 received < 11 needed
```

There are audio packet losses too, so it's not just video
```
00:00:26 - SDL Info (0): Returning RTP packet queued for too long
00:00:26 - SDL Info (0): Received OOS audio data (expected 2052, but got 2054)
00:00:26 - SDL Info (0): Returning RTP packet queued for too long
00:00:26 - SDL Info (0): Received OOS audio data (expected 2076, but got 2077)
00:00:26 - SDL Info (0): Returning RTP packet queued for too long
00:00:26 - SDL Info (0): Received OOS audio data (expected 2100, but got 2101)
00:00:26 - SDL Info (0): Returning RTP packet queued for too long
00:00:26 - SDL Info (0): Received OOS audio data (expected 2110, but got 2111)
```

The reason it "works" on Android (and only with H.264) is because Qualcomm SoCs on Android support H.264 reference picture invalidation. This allows the frame loss to be concealed with only minor artifacting until the encoder produces a frame that no longer references the dropped frame. You're still dropping frames, but they're less noticeable because it doesn't require a key frame each time a frame is dropped.

For whatever reason (maybe FFmpeg or hardware decoders themselves), decoders on PC don't support this. Enabling it on PC causes lingering artifacts after a frame loss happens, or even can cause the decoder to reject frames entirely and force a reset (which causes the entire video stream to blank out for a split second).

Long story short, at the moment this is working as expected. Your network connection is slightly too lossy for Moonlight to handle. Nvidia's official clients have a special command that can be sent to the host to raise and lower the FEC level depending on network conditions. Once I can reverse engineer that command format, I'll be able to implement dynamic FEC adjustments which would likely address this issue with your network connection (at least for the video-side, audio FEC is not yet understood by Moonlight).

**Comment by nk33** [comment]
How would i find out what is causing the packet loss?  I remember i used to play using PC a lot and never had this issue....its only now i have returned to it that this is happening

My whole house is connected via gigabit ethernet - all PC's and devices

why does the packet loss not occur when on same LAN but does through the router? is there maybe a port i am not forwearding that needs to be? have any ports been added to the port forward list in recent years that i havent added?

**Comment by nk33** [comment]
from those l;ogs it always seems to be 1 frame short - like it stops short 1 frame before finishing the job....is this router related or something ele in my network environment?

I notcied certain things like, for example, whatsapp videos - when i try and send via wifi from my home network they never send fully - have to keep retrying - yet from 4g/5g they just go non-stop until they are sent..

could these 2 situations be related?

**Comment by nk33** [comment]
After doing a bit of reading - it seems the Virgin Hub 3 is simply not good enough to process the packets. Obviously when it worked before i had much fewer devices connected to my LAN. Now we have everything from multiple pcs, tvs,, alexa, emby streaming etc...

I have an AX3600 i recently bought just as a wfi 6 AP - however i may put my virgin hub into modem mode and use this as a router when i get home.

will update you on the results
